activejob/lib/active_job/async_job.rb

```diff
@@ -0,0 +1,102 @@
+  #
+  # To use Async Job set the queue_adapter config to +:async+.
+  #
+  #   Rails.application.config.active_job.queue_adapter = :async
+  #
+  # Async Job supports job queues specified with +queue_as+. Queues
+  # are implemented as individual thread pools. Queues will be created
+  # automatically as needed using a default configuration. Should there be a
+  # need to customize the thread pool behind one or more queues, any
+  # concurrent-ruby thread pool can be injected. Simply create the thread
+  # pool directly then call the +create_queue+ method passing the thread
+  # pool as the second parameter:
+  #
+  #   thread_pool = Concurrent::FixedThreadPool.new(10, max_queue: 100, fallback_polcy: :caller_runs)
+  #   ActiveJob::AsyncJob.create_queue(:my_queue, thread_pool)
+  #
+  # The +create_queue+ method must be called during application initialization,
+  # before any jobs are post to the queue. Attempting to create a queue that
+  # already exists will raise an error.
+  class AsyncJob
+
+    DEFAULT_EXECUTOR_OPTIONS = {
+      min_threads:     [2, Concurrent.processor_count].max,
+      max_threads:     Concurrent.processor_count * 10,
+      auto_terminate:  true,
+      idletime:        60, # 1 minute
+      max_queue:       0, # unlimited
+      fallback_policy: :caller_runs # shouldn't matter -- 0 max queue
+    }.freeze
+
+    QUEUES = ThreadSafe::Cache.new do |hash, queue_name| #:nodoc:
+      hash[queue_name] = ActiveJob::AsyncJob.create_thread_pool
+    end
+
+    # Raised when an attempt is made to create a queue that already exists.
+    class QueueCreationError < ArgumentError; end
+
+    class << self
+      # Force all jobs to run synchronously when testing the Active Job gem.
+      # This should only be called from within unit tests.
+      def set_test_mode! #:nodoc:
+        @test_mode = true
+      end
+      # Allow jobs to run asynchronously when testing the Active Job gem.
+      # This should only be called from within unit tests.
+      def set_normal_mode! #:nodoc:
+        @test_mode = false
+      end
+
+      # Create a new job queue with the given +name+. Jobs will run on the
+      # given +thread pool+. The thread pool must be a concurrent-ruby
+      # {executor}[http://ruby-concurrency.github.io/concurrent-ruby/file.thread_pools.html].
+      # Raises +QueueCreationError+ when the queue already exists.
+      def create_queue(name, thread_pool)
+        raise QueueCreationError.new('queue already exists') if QUEUES.key? name
+        # possible race condition here but the use case is very narrow
+        QUEUES[name] = thread_pool
+      end
+
+      def create_thread_pool #:nodoc:
+        if @test_mode
+          Concurrent::ImmediateExecutor.new
+        else
+          Concurrent::ThreadPoolExecutor.new(DEFAULT_EXECUTOR_OPTIONS)
+        end
+      end
+
+      def enqueue(job_data, queue: 'default') #:nodoc:
+        QUEUES[queue].post(job_data) {|job| ActiveJob::Base.execute(job) }
```

Comment:
@jdantonio we need to use `compute_if_absent` up above:

```ruby
QUEUES = ThreadSafe::Cache.new do |hash, queue_name| #:nodoc:
  hash.compute_if_absent(queue_name) { ActiveJob::AsyncJob.create_thread_pool }
end
```

It is my fault there are no docs for `ThreadSafe::Cache` - there are no exclusivity/synchronization guarantees about the "default block" of `ThreadSafe::Cache.new`. The cache will happily let multiple threads inside.